### PR TITLE
drivers: sensors: Implement MAX31865 sensor

### PR DESCRIPTION
**Pull request by maass-hamburg**
This commit implements the temperature sensor interface for
the Maxim MAX31865 SPI Temperature Sensor. It also provides a sample.

Signed-off-by: Fin Maaß <fin.maass@haw-hamburg.de>

### CODE REVIEW

File: dts/bindings/sensor/maxim,max31865.yaml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        required: false
+        description: Low fault threshold (ADC CODE)
+
+  high-threshold:
+        type: int
+        default: 32767
+        required: false
+        description: High fault threshold (ADC CODE)
+
+  three-wire:
```
Comment (galak) [inline dts/bindings/sensor/maxim,max31865.yaml:40]: use `maxim,3-wire` as property name to match "upstream" linux binding:

https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/iio/temperature/maxim,max31865.yaml

**Review by galak — CHANGES_REQUESTED** [review]
* a fair amount of extra whitespace in driver needs cleaning up
* please add to `tests/drivers/build_all/sensor`
* minor fix for dts binding
* Do we need a unique sample for this sensor?
* only glanced over the driver code, is newlib truly needed for this driver?

**Comment by maass-hamburg** [comment]
> * Do we need a unique sample for this sensor?
> * only glanced over the driver code, is newlib truly needed for this driver?

Most sensor drivers have their own sample so this drive should also have its own.
The newlib is necessary for the `sqrt` funktion in the max31865.c to calculate the temperature.

File: drivers/sensor/max31865/Kconfig [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# MAX31865 temperature sensor configuration options
+
+# Copyright (c) 2022 HAW Hamburg FTZ-DIWIP
+# SPDX-License-Identifier: Apache-2.0
+
+config MAX31865
+	bool "MAX31865 Temperature Sensor"
+	depends on SPI && NEWLIB_LIBC
```
Comment (MaureenHelm) [inline drivers/sensor/max31865/Kconfig:8]:
```suggestion
	depends on NEWLIB_LIBC && DT_HAS_MAXIM_MAX31865_ENABLED
	select SPI
```

File: drivers/sensor/max31865/max31865.c [hunk 3]
```diff
@@ -0,0 +1,299 @@
+/*
+ * Copyright (c) 2022 HAW Hamburg FTZ-DIWIP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include "max31865.h"
+
+int8_t initMax31865(const struct device *dev)
```
Comment (MaureenHelm) [inline drivers/sensor/max31865/max31865.c:8]: Please no camel case

File: drivers/sensor/max31865/max31865.c [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+	max31865_set_vbias(dev, true);
+	union read_reg_u {
+		uint8_t u8[2];
+		uint16_t u16;
+	} read_reg;
+
+	read_reg.u16 = 0;
+	double resistance = 0;
+
+	k_sleep(K_MSEC(128));
```
Comment (MaureenHelm) [inline drivers/sensor/max31865/max31865.c:97]: Please document why this value was chosen

File: dts/bindings/sensor/maxim,max31865.yaml [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        type: int
+        default: 0
+        required: false
+        description: Low fault threshold (ADC CODE)
+
+  high-threshold:
+        type: int
+        default: 32767
+        required: false
+        description: High fault threshold (ADC CODE)
```
Comment (MaureenHelm) [inline dts/bindings/sensor/maxim,max31865.yaml:38]: Property defaults need justification in the description.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Please squash the requested changes into the original commit, and rebase to latest main to remove the merge commit.

**Comment by MaureenHelm** [comment]
> Please squash the requested changes into the original commit, and rebase to latest main to remove the merge commit.

This still needs to be done.

**Comment by maass-hamburg** [comment]
> > Please squash the requested changes into the original commit, and rebase to latest main to remove the merge commit.
> 
> This still needs to be done.

It should be done now. 
With this last commit I also removed the dependency on newlib by implementing a less accurate  function which is used, when newlib is not enabled.

**Comment by maass-hamburg** [comment]
> > Please squash the requested changes into the original commit, and rebase to latest main to remove the merge commit.
> 
> This still needs to be done.

It should be done now. This is my first commit to Zephyr and I am not that used to work with some git functions.

File: drivers/sensor/max31865/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,9 @@
+# Makefile - MAX31865
```
Comment (MaureenHelm) [inline drivers/sensor/max31865/CMakeLists.txt:1]: This isn't a Makefile

File: drivers/sensor/max31865/max31865.c [hunk 7]
```diff
@@ -0,0 +1,318 @@
+/*
+ * Copyright (c) 2022 HAW Hamburg FTZ-DIWIP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include "max31865.h"
+
+int initmax31865(const struct device *dev)
```
Comment (MaureenHelm) [inline drivers/sensor/max31865/max31865.c:8]: All of the functions in this file should be static
```suggestion
static int init_max31865(const struct device *dev)
```

File: drivers/sensor/max31865/max31865.c [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+{
+	const struct max31865_config *cfg = dev->config;
+
+	if (!spi_is_ready(&cfg->spi)) {
+		return -1;
```
Comment (MaureenHelm) [inline drivers/sensor/max31865/max31865.c:13]:
```suggestion
		return -ENODEV;
```

File: drivers/sensor/max31865/max31865.c [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+		  .len = len }
+	};
+	const struct spi_buf_set rx = {
+		.buffers = rx_buf,
+		.count = 2
+	};
+
+	ret = spi_transceive_dt(&cfg->spi, &tx, &rx);
+	if (ret) {
+		LOG_DBG("spi_transceive FAIL %d\n", ret);
```
Comment (MaureenHelm) [inline drivers/sensor/max31865/max31865.c:71]:
```suggestion
		LOG_ERR("spi_transceive FAIL %d\n", ret);
```

File: drivers/sensor/max31865/max31865.h [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+	bool filter_50hz;               /**< Filter_50hz 50 Hz filter enabled (@a true) or 60 Hz filter enabled (@a false). */
+	uint16_t low_threshold;         /**< Low fault threshold. */
+	uint16_t high_threshold;        /**< High fault threshold. */
+};
+
+/* Bit manipulation macros */
+#define SETBIT(x, y)        (x) |= BIT(y)
+#define CLEARBIT(x, y)      (x) &= ~BIT(y)
+#define TESTBIT(data, pos)  ((0u == (data & BIT(pos)))?0u:1u)
+#define SWAP(a, b) (((a) ^= (b)), ((b) ^= (a)), ((a) ^= (b)))
```
Comment (MaureenHelm) [inline drivers/sensor/max31865/max31865.h:95]: Please use macros from `include/zephyr/sys/util*.h` instead

File: drivers/sensor/max31865/max31865.c [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+	SWAP(read_reg.u8[0], read_reg.u8[1]);
+
+	LOG_DBG("RAW: %02X %02X , %04X", read_reg.u8[0], read_reg.u8[1], read_reg.u16);
+	if (TESTBIT(read_reg.u16, 0)) {
+		uint8_t fault_register;
+		max31865_spi_read(dev, (REG_FAULT_STATUS), &fault_register, 1);
+		/*Clear fault register */
+		data->config_control_bits |= 0x02;
+		configure_device(dev);
+		LOG_WRN("Fehler: %10x", fault_register);
```
Comment (MaureenHelm) [inline drivers/sensor/max31865/max31865.c:114]: English please

File: drivers/sensor/max31865/max31865.c [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+	/* Set the PTD. */
+	data->resistance_at_zero = config->resistance_at_zero;
+	data->resistance_reference = config->resistance_reference;
+	/* Send configuration to device */
+	configure_device(dev);
+	/* Send threshold to device */
+	set_threshold_values(dev);
+}
+
+int max31865_getTemperatureSensor(const struct device *dev, struct sensor_value *val)
```
Comment (MaureenHelm) [inline drivers/sensor/max31865/max31865.c:242]:
```suggestion
static int max31865_get_temperature_sensor(const struct device *dev, struct sensor_value *val)
```

File: drivers/sensor/max31865/max31865.h [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+
+void max31865_init(const struct device *dev);
+void max31865_set_vbias(const struct device *dev, bool enable);
+int max31865_get_temperature(const struct device *dev, double *temperature);
+int max31865_spi_write(const struct device *dev, uint8_t reg, uint8_t *data, uint8_t lenght);
+int initmax31865(const struct device *dev);
+int max31865_getTemperatureSensor(const struct device *dev, struct sensor_value *val);
+void configure_device(const struct device *dev);
+void set_threshold_values(const struct device *dev);
+double calculate_temperature(double resistance, double resistance_0);
```
Comment (MaureenHelm) [inline drivers/sensor/max31865/max31865.h:105]: Remove

File: dts/bindings/sensor/maxim,max31865.yaml [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+        type: int
+        default: 100
+        required: false
+        description: Sensor resistance at 0 Celsius (100 for PT100, 1000 for PT1000)
+
+  resistance-reference:
+        type: int
+        default: 430
+        required: false
+        description: Circuit reference resistance (recommended on MAX31865 datasheet 400 ohms for PT100, 4000 for PT1000, the Adafruit boards use 430 ohms and 4300 ohms)
```
Comment (MaureenHelm) [inline dts/bindings/sensor/maxim,max31865.yaml:26]: These should be required, since we can't assume what is on the board.

File: samples/sensor/max31865/README [hunk 15]
```diff
@@ -0,0 +1,3 @@
+This is a sample app to read a MAX31865.
```
Comment (MaureenHelm) [inline samples/sensor/max31865/README:1]: This file should be in .rst format. See `samples/hello_world/README.rst` for example

File: samples/sensor/max31865/CMakeLists.txt [hunk 16]
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.20.0)
```
Comment (MaureenHelm) [inline samples/sensor/max31865/CMakeLists.txt:3]: Please move the sample changes to a separate commit in this PR. Please also add a `sample.yaml`

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: samples/sensor/max31865/prj.conf [hunk 17]
```diff
@@ -0,0 +1,11 @@
+CONFIG_PRINTK=y
+CONFIG_CONSOLE=y
+CONFIG_USE_SEGGER_RTT=y
+CONFIG_RTT_CONSOLE=y
+CONFIG_UART_CONSOLE=n
+CONFIG_SHELL=y
+CONFIG_SHELL_BACKEND_RTT=y
+CONFIG_SERIAL=n
```
Comment (dcpleung) [inline samples/sensor/max31865/prj.conf:8]: Does this require shell to work properly? If not, please remove `CONFIG_SHELL`.

Also, This looks like a generic sample where any board with a SPI bus can be used. So I think it is better to move the config to enabling RTT and disabling UART to a board overlay (under this sample). This will allow other boards with different serial/console to use this sample.

File: samples/sensor/max31865/sample.yaml [hunk 18]
```diff
@@ -0,0 +1,10 @@
+sample:
+  name: MAX31865 temperature sensor
+tests:
+  sample.sensor.max31865:
+    harness: sensor
+    tags: samples sensors
+    depends_on: spi
+    platform_allow: particle_xenon
```
Comment (dcpleung) [inline samples/sensor/max31865/sample.yaml:8]: Remove the `platform_allow` so this is not restricted to one board.

And add `filter: dt_compat_enabled("maxim,max31865")` so this can be built (via twister) any boards.

File: drivers/sensor/max31865/max31865.c [hunk 19]
```diff
@@ -0,0 +1,302 @@
...
+		(config->high_threshold >> 7) & 0x00ff, (config->high_threshold << 1) & 0x00ff,
+		(config->low_threshold >> 7) & 0x00ff, (config->low_threshold << 1) & 0x00ff};
+	int err = max31865_spi_write(dev, WR(REG_HIGH_FAULT_THR_MSB), cmd, 4);
+
+	if (err < 0) {
+		LOG_ERR("Error write SPI%d\n", err);
+	}
+}
+
+#ifdef NEWLIB_LIBC
```
Comment (dcpleung) [inline drivers/sensor/max31865/max31865.c:149]: This needs to be `CONFIG_NEWLIB_LIBC`.

**Review by dcpleung — CHANGES_REQUESTED** [review]
BTW, please reverse the order of the commit so the introduction of the driver comes first.

File: samples/sensor/max31865/src/main.c [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+	union sensor_val sv;
+	int r;
+
+	while (1) {
+
+		r = sensor_sample_fetch(max_sensor);
+		if (r != 0) {
+			continue;
+		}
+		sensor_channel_get(max_sensor, SENSOR_CHAN_AMBIENT_TEMP, &sv.str);
```
Comment (gmarull) [inline samples/sensor/max31865/src/main.c:31]: let's not add yet another sensor sample without being generic for the sensor type (unless it's really needed to demonstrate special features). See the work done for the generic accelerometer polling sample.

**Review by gmarull — CHANGES_REQUESTED** [review]

File: samples/sensor/max31865/src/main.c (hunk 20 above)
Comment (maass-hamburg) [reply samples/sensor/max31865/src/main.c:31]: So should the sample be removed?

**Comment by maass-hamburg** [comment]
I removed the sample. It is still available in another branch [https://github.com/FTZ-DiWiP/zephyr/tree/max31865_sample](url) and could be added later, if wished. But this pull request now only provides the driver without the sample.

File: drivers/sensor/max31865/max31865.h [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+
+#include <math.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/gpio.h>
```
Comment (MaureenHelm) [inline drivers/sensor/max31865/max31865.h:19]: Is this needed? I don't see any GPIO operations in the driver

File: drivers/sensor/max31865/max31865.h [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+#include <zephyr/sys/util_macro.h>
+#include <zephyr/kernel.h>
+LOG_MODULE_REGISTER(MAX31865, CONFIG_SENSOR_LOG_LEVEL);
+
+#define MAX31865_FAULT_HIGH_THRESHOLD (1 << 7)
+#define MAX31865_FAULT_LOW_THRESHOLD  (1 << 6)
+#define MAX31865_FAULT_REFIN	      (1 << 5)
+#define MAX31865_FAULT_REFIN_FORCE    (1 << 4)
+#define MAX31865_FAULT_RTDIN_FORCE    (1 << 3)
+#define MAX31865_FAULT_VOLTAGE	      (1 << 2)
```
Comment (MaureenHelm) [inline drivers/sensor/max31865/max31865.h:32]: Please use `BIT()`

File: drivers/sensor/max31865/max31865.h [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+ * is specified below.
+ */
+#define RTD_A (RTD_A_ITS90)
+#define RTD_B (RTD_B_ITS90)
+
+struct max31865_data {
+	struct sensor_value val;
+	uint8_t config_control_bits;
+	uint16_t resistance_at_zero;
+	uint16_t resistance_reference;
```
Comment (MaureenHelm) [inline drivers/sensor/max31865/max31865.h:76]: None of these appear to ever change at runtime, so there's no reason to keep copies in this struct. Just access them from the config struct directly

File: dts/bindings/sensor/maxim,max31865.yaml [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+
+  resistance-at-zero:
+        type: int
+        required: true
+        description: Sensor resistance at 0 Celsius (100 for PT100, 1000 for PT1000)
+
+  resistance-reference:
+        type: int
+        required: true
+        description: Circuit reference resistance (recommended on MAX31865 datasheet 400 ohms for PT100, 4000 for PT1000, the Adafruit boards use 430 ohms and 4300 ohms)
```
Comment (MaureenHelm) [inline dts/bindings/sensor/maxim,max31865.yaml:24]: Units?

File: dts/bindings/sensor/maxim,max31865.yaml [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+
+  low-threshold:
+        type: int
+        default: 0
+        required: false
+        description: Low fault threshold (ADC CODE, 15-bit value, default value is the minimum value)
```
Comment (MaureenHelm) [inline dts/bindings/sensor/maxim,max31865.yaml:30]: Units?

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/sensor/max31865/max31865.h (hunk 23 above)
Comment (maass-hamburg) [reply drivers/sensor/max31865/max31865.h]: `config_control_bits` get changed at runtime, but the others don't.

**Comment by maass-hamburg** [comment]
So any other requested changes?

File: drivers/sensor/max31865/max31865.c [hunk 26]
```diff
@@ -0,0 +1,304 @@
...
+	/* Read resistance measured value */
+	int err = max31865_spi_read(dev, (REG_RTD_MSB), read_reg.u8, 2);
+
+	max31865_set_vbias(dev, false);
+
+	if (err < 0) {
+		LOG_ERR("SPI read %d\n", err);
+		return -EIO;
+	}
+	byteswp(read_reg.u8, read_reg.u8 + 1, 1);
```
Comment (MaureenHelm) [inline drivers/sensor/max31865/max31865.c:212]: I think you want `sys_be16_to_cpu` instead

File: drivers/sensor/max31865/max31865.c [hunk 27]
```diff
@@ -0,0 +1,304 @@
...
+		.conversion_mode = false,                                                          \
+		.one_shot = true,                                                                  \
+		.three_wire = DT_INST_PROP(inst, maxim_3_wire),                                    \
+		.fault_cycle = MAX31865_FAULT_DETECTION_NONE,                                      \
+		.filter_50hz = DT_INST_PROP(inst, filter_50hz),                                    \
+		.low_threshold = DT_INST_PROP(inst, low_threshold),                                \
+		.high_threshold = DT_INST_PROP(inst, high_threshold),                              \
+	};                                                                                         \
+                                                                                                   \
+	DEVICE_DT_INST_DEFINE(inst, max31865_init, NULL, &max31865_data_##inst,                    \
```
Comment (MaureenHelm) [inline drivers/sensor/max31865/max31865.c:299]: Use `SENSOR_DEVICE_DT_INST_DEFINE` instead

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
